subsys/bluetooth/audio/mcc.c

```diff
@@ -1195,19 +1182,283 @@ static uint8_t discover_include_func(struct bt_conn *conn,
 }
 #endif /* CONFIG_BT_MCC_OTS */
 
+/* Start discovery of included services */
+static void discover_included(struct bt_conn *conn)
+{
+	int err;
+
+	discover_params.start_handle = cur_mcs_inst->start_handle;
+	discover_params.end_handle = cur_mcs_inst->end_handle;
+	discover_params.type = BT_GATT_DISCOVER_INCLUDE;
+	discover_params.func = discover_include_func;
 
-/* This function is called when characteristics are found.
- * The function will store handles, and optionally subscribe to, GMCS
- * characteristics.
+	BT_DBG("Start discovery of included services");
+	err = bt_gatt_discover(conn, &discover_params);
+	if (err) {
+		discovery_failure(conn, "Discover of included service failed", err);
+	}
+}
+
+static void subscribe_mcs_char_func(struct bt_conn *conn);
+
+static void subscription_cb(struct bt_conn *conn, uint8_t err,
+			    struct bt_gatt_subscribe_params *params)
+{
+	if (err) {
+		discovery_failure(conn, "Subscription callback error", err);
+	}
+
+	BT_DBG("Subscribed: value handle: %d, ccc handle: %d",
+	       params->value_handle, params->ccc_handle);
+
+	/* Subscribe to next characteristic */
+	subscribe_mcs_char_func(conn);
+}
+
+/* Subscribe to a characteristic - helper function */
+static int do_subscribe(struct bt_conn *conn, uint16_t handle,
+			 struct bt_gatt_subscribe_params *sub_params)
+{
+	/* With ccc_handle == 0 it will use auto discovery */
+	sub_params->ccc_handle = 0;
+	sub_params->end_handle = cur_mcs_inst->end_handle;
+	sub_params->value = BT_GATT_CCC_NOTIFY;
+	sub_params->value_handle = handle;
+	sub_params->notify = mcs_notify_handler;
+	sub_params->subscribe = subscription_cb;
+
```

Comment:
That would seem natural, yes.  And it would save some lines of code.

The reason I like to have it where it currently is is that that pointer is also used as the flag for whether the characteristic has already been subscribed to or not.  Having the "setting" of the flag close to the testing of the flags helps understanding, IMO.
But a net saving of 16 lines of code trumps that, I guess.  Will change.

Reply:
I get that, but you are also checking the handle which are set by this function, so the argument isn't really solid :)